File: packages/sdk-ts/src/resources/feedback.ts
import type { Result } from "../errors.js";
import { success, failure } from "../errors.js";

export interface FeedbackFilters {
  template?: string;
  outcome?: string;
  limit?: number;
  offset?: number;
}

export class FeedbackResource {
  constructor(
    private baseUrl: string,
    private headers: () => Record<string, string>,
  ) {}

  async query(filters?: FeedbackFilters): Promise<Result<Record<string, unknown>>> {
    const params = new URLSearchParams();
    if (filters?.template) params.set("template", filters.template);
    if (filters?.outcome) params.set("outcome", filters.outcome);
    if (filters?.limit !== undefined) params.set("limit", String(filters.limit));
    if (filters?.offset !== undefined) params.set("offset", String(filters.offset));

    const qs = params.toString();
    const url = `${this.baseUrl}/api/v1/feedback${qs ? `?${qs}` : ""}`;

    try {
      const res = await fetch(url, { method: "GET", headers: this.headers() });
      const body = await res.json();

      if (!res.ok) {
        const apiError = body.error || {};
        return {
          data: null,
          error: {
            type: apiError.type || "api_error",
            code: apiError.code || "unknown",
            message: apiError.message || `Request failed with status ${res.status}`,
            statusCode: res.status,
          },
        };
      }

      return success(body);
    } catch (err) {
      return {
        data: null,
        error: {
          type: "network_error",
          code: "network_error",
          message: err instanceof Error ? err.message : "Network error",
          statusCode: 0,
        },
      };
    }
  }
}

File: packages/sdk-py/gatewerk/resources/feedback.py
"""Feedback resource for the Gatewerk Python SDK."""

from __future__ import annotations

from typing import Any, Optional

from .._base import BaseResource
from ..types import FeedbackList


class FeedbackResource(BaseResource):

    def query(
        self,
        *,
        template: Optional[str] = None,
        outcome: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        timeout: Optional[float] = None,
    ) -> FeedbackList:
        params: dict[str, Any] = {}
        if template is not None:
            params["template"] = template
        if outcome is not None:
            params["outcome"] = outcome
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        data = self._request("GET", "/api/v1/feedback", params=params, timeout=timeout)
        return FeedbackList.model_validate(data)
